library(tidyverse)
library(seqinr)

ship_feat<-read_tsv("../Starships/metadata/ships/starfish/output/mycodb.final.starships.feat")

############
# manual set
############
# read in fasta and taxonomic assignments
man_fa<-read.alignment("../Starships/metadata/ships/manual/Starships.fa",format="fasta")
man_tax<-read_csv("../Starships/metadata/ships/manual/Starships.fulltaxa.csv")

# extract sequences and headers from multifasta
man_ship_table<-map_df(1:length(man_fa$nam),~{
  id<-man_fa$nam[.x]
  seq<-gsub("-","",man_fa$seq[.x])
  spec<-man_tax %>% filter(Code == id) %>% pull(Species)
  tibble(genome_name=id,genome_sequence=seq,organism=spec)
}) 

##############
# starfish set
##############
# read in fasta and taxonomic assignments
star_fa<-read.alignment("../Starships/metadata/ships/starfish/output/mycodb.final.starships.fna",format="fasta")
star_tax<-read_tsv("../Starships/MTDB/mycodb2899.ome2species.txt",col_names=c("ome","species"))

# extract sequences and headers from multifasta
# star_ship_table<-map_df(1:length(star_fa$nam),~{
#   id<-gsub("|+","",gsub("|-","",star_fa$nam[.x],fixed=TRUE),fixed=TRUE)
#   ome_id<-str_split(id,"_",simplify=TRUE)[1]
#   seq<-gsub("-","",star_fa$seq[.x])
#   spec<-star_tax %>% filter(ome == ome_id) %>% pull(species)
#   tibble(genome_name=id,genome_sequence=seq,organism=spec)
# }) 

# * now working with proper gff and faa files from emile
# TODO: check to see if ome prefixes in headers represent unique ships. if not, then need to split faa but gene names for each ship in gff (after adding sequence ID to gff)
"mycodb.ogCargo.aug.empty.faa"
"mycodb.ogCargo.aug.empty.gff"

##############
# genome table
##############
# output csv with columns: id, genome_name, genome_sequence, organism

both_df<-bind_rows(man_ship_table,star_ship_table) %>% mutate(id=row_number()+1) %>% relocate(id) 
both_df %>%
  write_csv("../Starships/metadata/ships/all-ship-mas-input.csv")

#######################
# feature and ann table
#######################
# list all files from generated by diamond blast
blast_files<-Sys.glob("../Starships/*/*/blast/*.txt")
gene_df<-map_df(blast_files,~{
  blast.res<-read_tsv(gsub(".faa",".txt",.x),show_col_types=FALSE,col_names=c("qseqid","sseqid","pident","length","mismatch","gapopen","qstart","qend","sstart","send","evalue","bitscore","qseq","qseq_translated"))
  split.name<-str_split(basename(.x),"\\.",simplify=TRUE)
  gene <- split.name[length(split.name) - 1]
  blast.res %>% mutate(annotation=gene) %>% mutate(across(everything(),~as.character(.)))
})

# join with already existing gene annotations
exist_gene_files<-c(Sys.glob("../Starships/captain/*/faa/starfish/*.faa"),Sys.glob("../Starships/captain/*/faa/manual/*.faa"))
exist_genes<-map_df(exist_gene_files,~{
  ship_name<-gsub(".*part_|.faa|-Captain","",basename(.x))
  if (ship_name %in% ship_feat$captainID) {
    split.name<-str_split(basename(.x),"\\.",simplify=TRUE)
    gene_id<-gsub(".*part_|.faa|-Captain","",basename(.x))
    gene <- split.name[length(split.name) - 1]
    x.fasta<-read.alignment(.x,format="fasta")
    header<-x.fasta$nam[[1]]
    seq<-x.fasta$seq[[1]]
    tibble(gene_id=gene_id,annotation=gene,header=header,sequence=seq,public_notes=NA,private_notes=NA,flag=0,assigned_to_id=1)
  }
})

########
# combine with existing genes
# rate the quality of blast hits
ann_df<-gene_df %>%
  mutate(public_notes=NA,private_notes=NA,assigned_to_id=1,
    flag= case_when(
        pident < 50 ~ "2",
        pident >= 50 & pident < 85 ~ "1",
        pident >= 85 ~ "0"
      )) %>%
  mutate(across(everything(),~as.character(.))) %>%
  dplyr::rename(sequence="qseq_translated") %>%
  bind_rows(exist_genes %>% select(annotation, sequence, public_notes,	private_notes,	flag,	assigned_to_id) %>% mutate(across(everything(),~as.character(.)))) %>%
  mutate(id=row_number()) %>%
  select(id, annotation, sequence, public_notes,	private_notes,	flag,	assigned_to_id) #%>% filter(!is.na(sequence))
# export csv with columns: id	annotation	`sequence`	public_notes	private_notes	flag	assigned_to_id
ann_df %>%
  write_csv("../Starships/metadata/ships/all-ann-mas-input.csv",quote="needed",na="")

########
# TODO: join faa with bed features using gene_id or faa header?
star_bed<-read_tsv("../Starships/MTDB/mycodb.final.starships.bed",col_names=c("chr","start","end","gene_id","feature","strand","starshipID","attribute"))
ars_bed <- read_tsv("../arsenic/ogCargo5_arsenic7.f20000.bed",col_names=c("chr","start","end","gene_id","feature","strand","starshipID","note")) %>%
  mutate(across(c(start,end),~as.numeric(.)))

# ars_inner<-gene_df %>% inner_join(ars_bed,by="gene_id") %>% pull(gene_id)
# star_inner<-gene_df %>% inner_join(star_bed,by="gene_id")%>% pull(gene_id)
# intersect(ars_inner,star_inner) %in% star_inner
# * ars_bed has all the annotations we need

# create manual bed file
# FIXME: assuming hits to positive strand only...
man_bed <- gene_df %>% filter(qseqid %in% man_ship_table$genome_name) %>% 
  mutate(chr=qseqid,start=qstart,end=qend,feature=annotation,strand="+",starshipID=qseqid,attribute=annotation,
    across(c(start,end),~as.numeric(.)))

both_bed<-full_join(man_bed,ars_bed) %>% 
  left_join(ship_feat,by=c("chr"="#contigID","starshipID"),suffix=c(".bed",".feat")) %>%
  rowwise() %>%
  mutate(across(c(start,end),~abs(elementBegin-.))) %>% ungroup()

# export csv with columns: id	`start`	stop	`type`	strand	annotation_id	genome_id
feat_df<-both_bed  %>%
  inner_join(gene_df %>% mutate(annotation_id=row_number()),by=c("qseqid","feature"="annotation"),suffix=c(".gene",".ann")) %>%
  # mutate(feature=fct_recode(feature,"tyr"="cap")) %>%
  inner_join(both_df,by=c("starshipID"="genome_name")) %>% 
  mutate(start=qstart.gene,stop=qend.gene,strand=strand.bed) %>%
  dplyr::rename(genome_id="id",type="feature") %>%
  select(start, stop, type,	strand,	annotation_id,	genome_id) %>%
  mutate(id=row_number())

feat_df %>%
  write_csv("../Starships/metadata/ships/all-feat-mas-input.csv",quote="needed",na="")
